254 check translations

mrsarm opened the pull request:
Ticket: https://github.com/medic/medic-conf/issues/254

### Check translations content before push

- Add "messageformat" module to check translations that use that syntax
- If a translation use the "MessageFormat" syntax check the translation and fails the process if there is an error
- If a translation is empty just warn the the lack of the translation

#### Warnings

This is how warnings are displayed when empty messages are found (the upload process is not interrupted):

```
$ medic-conf upload-custom-translations --local
INFO Using local url from COUCH_URL environment variable: http://admin:****@localhost:5984/medic 
INFO Processing config in default. 
INFO Actions:
     - upload-custom-translations
INFO Starting action: upload-custom-translations…
WARN Empty 'en' translation for 'report.delivery.babys_condition.baby_repeat' key
WARN Empty 'en' translation for 'report.delivery.pnc_visits.who_note' key
...
```

#### Errors

MessageFormat wrong syntax does raise errors and exit the process:

```
$ medic-conf upload-custom-translations --local
INFO Using local url from COUCH_URL environment variable: http://admin:****@localhost:5984/medic
INFO Processing config in default.
INFO Actions:
     - upload-custom-translations
INFO Starting action: upload-custom-translations…
ERROR Cannot compile 'en' translation n.month = '{MONTHS, plural one{1 month} other{# months}}' : Expected "," but "o" found.
ERROR Cannot compile 'en' translation n.day = '{DAYS, plural, one{1 day} other{# days}' : Expected "=", "}", or identifier but end of input found.
ERROR There were 2 errors trying to compile the translations
```

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +89,34 @@ function parse(filePath, options) {
+      try {
+        mf.compile(msgSrc);
+      } catch (e) {
+        if (!HAS_MUSTACHE_MATCHER.test(msgSrc)) {
```
Nitpick: Is this because mustache format breaks message format? If so, I'd run this test before the compilation step, because there's no point running compile if we're going to ignore any errors, right?

mrsarm replied:
> Is this because mustache format breaks message format?

Yes, and checking some repos with translations, we don't mix both systems mustache and messageformat in a single translation.

> there's no point running compile if we're going to ignore any errors, right?

But not all errors are ignored, only if there is a mustache expression.

I can change the code to this, but still looks cumbersome:

```js
    if (!msgSrc) {
      log.warn(`Empty '${languageCode}' translation for '${msgKey}' key`);
    } else if (mf && !HAS_MUSTACHE_MATCHER.test(msgSrc)) {
      try {
        mf.compile(msgSrc);
      } catch (e) {
        log.error(`Cannot compile '${languageCode}' translation ${msgKey} = '${msgSrc}' : ${e.message}`);
        errorsFound++;
      }
    }
```

garethbowen replied:
That way looks better to me, but whatever you prefer is fine.

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +89,34 @@ function parse(filePath, options) {
+  let errorsFound = 0;
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    if (!msgSrc) {
+      log.warn(`Empty '${languageCode}' translation for '${msgKey}' key`);
```
I just realised this will spam the console for projects that have an empty language (eg: one they're still working on). Maybe keep a count of `emptyTranslations` and do one warning at the end with `${emptyTranslations} empty translations found for ${languageCode}`?

mrsarm replied:
Yes, taking into account that it is easy to find yourself the empty ones with an editor, it is a good idea.

mrsarm commented:
Hey @garethbowen , I changed some part of the codes with your suggestions, and added the placeholders check, taking into account the `messages-en.properties` as a base template like the "poe" CLI, and also it takes as a template a `messages-ex.properties` if it is found in the same folder.

If one error is found because the user mistype one placeholder key, it will see in the log something like:

```
ERROR Cannot compile 'hi' translation with key 'messages.d.validation.days_since_birth' has placeholders that do not match those of messages-en.properties                                                                                                                                                   
You can use messages-ex.properties to add placeholders missing from the reference context.                                                                  
ERROR There was 1 error trying to compile 'hi' translations
```

The problem is that most config repos I saw don't include the `-ex` file in their repo, so users will need to add it if they are using a placeholder that is not in the `-en` file, which makes sense but not sure how big is the impact of asking for that. I can also add in the log error a more clear message about that, but I wanted to confirm we are doing it right or maybe we can take into account `-ex` placeholders in another way.

Moreover, the translations "references" under the [cht-core#config/standard/translations](https://github.com/medic/cht-core/tree/master/config/standard/translations) folder won't pass the validations, I found a typo error [here](https://github.com/medic/cht-core/blob/master/config/standard/translations/messages-hi.properties#L15), note the uppercase `P` at `{{Patient_id}}` that should be lowercase, I'll provide a fix to that typo in the same PR where I'll add the same "messageformat" validations to the POE scripts.

Also I included a validations that if placeholders are used but the keys are not found in the `-en` or `-ex` files, it raises errors highlighting the missed keys, so running these validations under the same folder will raise this errors due missed keys in the `-en` file:

```
ERROR Cannot compile 'sw' translation with key 'messages.d.validation.delivery_code' has placeholders, but the key does not match any of messages-en.properties                                                                                                                                                         
You can use messages-ex.properties to add placeholders missing from the reference context.                                                                  
ERROR Cannot compile 'sw' translation with key 'messages.c.validation.patient_name' has placeholders, but the key does not match any of messages-en.properties                                                                                                                                                          
You can use messages-ex.properties to add placeholders missing from the reference context.                                                                  
ERROR Cannot compile 'sw' translation with key 'messages.n.validation.patient_name' has placeholders, but the key does not match any of messages-en.properties                                                                                                                                                          
You can use messages-ex.properties to add placeholders missing from the reference context.                                                                  
ERROR There were 3 errors trying to compile 'sw' translations
```

I can also add in the "poe" PR the fixes for these keys.

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
+    log.warn(`Cannot check '${lang}' translations: ${e.message}`);
+  }
+  const placeholders = extractPlaceholdersFromTranslations(translations);
+  let formatErrorsFound = 0,  placeholderErrorsFound = 0, emptiesFound = 0;
```
Nit: Put each variable on its own line, eg

```suggestion
  let formatErrorsFound = 0;
  let placeholderErrorsFound = 0;
  let emptiesFound = 0;
```

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    if (!msgSrc) {
+      emptiesFound++;
+    } else if (/{{[\s\w.#^/'|]+}}/.test(msgSrc)) {
```
Pull this regex out as a constant at the top of the file so it can be named for readability.

mrsarm replied:
Done

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
+              `has placeholders, but the key does not match any of ${EN_FILE}\n` +
+              'You can use messages-ex.properties to add placeholders missing from the reference context.');
+            placeholderErrorsFound++;
+          } else {
+            const foundAllPlaceholders = placeholder.every(el => templatePlaceholder.includes(el));
+            if (!foundAllPlaceholders) {
+              log.error(`Cannot compile '${lang}' translation with key '${msgKey}' ` +
+                `has placeholders that do not match those of ${EN_FILE}\n` +
+                'You can use messages-ex.properties to add placeholders missing from the reference context.');
+              placeholderErrorsFound++;
+            }
```
I think we should allow this case. If the English file defines some placeholders but the project chooses not to use them that's ok. So I think you can remove this whole block.

mrsarm replied:
No, it is the opposite, if a non-english translation don't use one or all of the placeholders used in the `-en` translation is fine, this block detects when a non-english translation uses a placeholder that is not present in the `-en` version (or `-ex`).

Eg. this is OK:

**messages-en.properties**
```
Number\ in\ month = {{count}} in {{month}}
```

**messages-es.properties**
```
Number\ in\ month = {{count}} total
```

This is WRONG:

**messages-en.properties**
```
Number\ in\ month = {{count}} in {{month}}
```

**messages-es.properties**
```
Number\ in\ month = {{count}} en {{month}} de {{another_placeholder}}
```

`{{another_placeholder}}` needs to be defined in messages-en.properties or messages-ex.properties first.

garethbowen replied:
Ahh sorry! I read that wrong.

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
+  // Extract from github.com/medic/cht-core/blob/master/scripts/poe/lib/utils.js
+  const result = {};
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    let placeholders = typeof msgSrc === 'string' ? msgSrc.match(/{{[\s\w.#^/'|]+}}/g) : null;
```
This regex is identical to the one above, so when you pull that out as a constant you can use it here too.

mrsarm replied:
Yes note that this one has the `/g` suffix at the end to make it "global" while the other not, but thinking about it again I can use the global expression there too.

mrsarm replied:
Done.

mrsarm replied:
@garethbowen I had to rollback the regex constant refactor, turns out regex/g are mutable objects :facepalm: : 

```
> // Not using constant the result is predictable:
undefined
> /{{[\s\w.#^/'|]+}}/g.test('Hey {{you}}!')
true
> /{{[\s\w.#^/'|]+}}/g.test('Hey {{you}}!')
true
> // Now using a constant the result change on each execution:
undefined
> const MUSTACHE_MATCHER = /{{[\s\w.#^/'|]+}}/g
undefined
> MUSTACHE_MATCHER.test('Hey {{you}}!')
true
> MUSTACHE_MATCHER.test('Hey {{you}}!')
false
> MUSTACHE_MATCHER.test('Hey {{you}}!')
true
```

Replacing the `/g` at the end of the regex by just `/` solve the problem with the `test(...)` method, but is needed by the `match(...)` method used in another part of the code. I can also remove the `/g` and use `matchAll` instead, but you need Node 12+.

garethbowen replied:
Wow, nice find!

I think you can use `match` and check for `null` instead of using `test`, eg:

```
> 'Hey {{you}}!'.match(MUSTACHE_MATCHER);
["{{you}}"]
> 'Hey '.match(MUSTACHE_MATCHER);
null
```

mrsarm replied:
Moved again as constant but avoiding to use the `test()` method that mutes the regex constant state

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
+      placeholders = placeholders
+        .sort()
+        .concat(extraPlaceholders[msgKey] ? extraPlaceholders[msgKey] : [])
+        .filter((el, i, a) => i === a.indexOf(el));
```
If I'm reading this right this is removing duplicate placeholders. It'd be more readable to either use lodash `_.uniq` or use a Set.

mrsarm replied:
Yes, copy&paste from "poe", but due that I'll change the code there, I'll also add the changes there.

mrsarm replied:
Done.

garethbowen commented on src/fn/upload-custom-translations.js:
```diff
@@ -81,6 +115,75 @@ function parse(filePath, options) {
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    let placeholders = typeof msgSrc === 'string' ? msgSrc.match(/{{[\s\w.#^/'|]+}}/g) : null;
+    if (placeholders) {
+      const msgExtraPlaceholders = extraPlaceholders[msgKey] ? extraPlaceholders[msgKey] : [];
+      result[msgKey] = _.uniq(placeholders.concat(msgExtraPlaceholders));
```
I think now that you're using lodash this can be made simpler...
```suggestion
      result[msgKey] = _.uniq(placeholders.concat(extraPlaceholders[msgKey] || []));
```

What do you think?

mrsarm commented:
@garethbowen I did address all the changes suggested. Do you want me to move this to a new Npm module as discussed last time? so we can reuse the same logic in the "poe" script.

In that case, I'm wondering what would be the best interface, the current "in place" implementation does the logging of warnings and errors, and in case of errors it aborts the process. Now we should create a method in the new module that receives as parameters the input options as it is now, but it should returns the validations as a result, so the consumer of the module decides what to do with the errors detected: log them as warns or errors, abort the process, ask for confirmation to continue...

**Here is my proposal:**

Module name: **_messages-checker_** (open to use another name, I'm really bad to name stuff),

Public method:

```js
/**
 * Check all the languages translations, optionally using `baseTranslations`
 * as template translations when placeholders are used.
 *
 * @param {string} dir - directory where the translation files with name messages-XX.properties are stored
 * @param {string[]} langs - the lowercase ISO 639-1 code languages to be checked. Set `null` to check all
 *        the languages found at `dir`
 * @param {string[]} baseTranslations - the ISO 639-1 code of the languages used as base template
 *        when placeholders are used
 *
 * @returns {Promise} resolved promise if there are no errors, or a rejected one with an array
 *                    of the validations failed. Eg.:
 *
 * [
 *   {"type":"missed-placeholder",  "lang": "es", "key":"n.month", "error":"key not found in base translations"},
 *   {"type":"wrong-placeholder",   "lang": "hi", "key":"user.greeting", "error":"placeholders not found in base translations"},
 *   {"type":"wrong-messageformat", "lang": "id", "key":"records.count", "error":"Expected \",\" but \"o\" found"},
 *   ...
 * ]
 */
async function checkTranslations(dir, langs, baseTranslations) {...};
```

It is just a proposal, open to as many corrections as needed.

garethbowen commented:
I've approved this PR because the code solid.

> Do you want me to move this to a new npm module as discussed last time?

Yeah let's do it if you feel comfortable doing it. I'm also happy for this to be ATed as is, and we can pull out a module another time, if you prefer.

> Module name: messages-checker (open to use another name, I'm really bad to name stuff),

Naming is hard. How about `translation-checker`?

> Public method:

I'm not sure why the `baseTranslations` are in a properties file - I think it would be much easier to work with if they were an array of known properties for each key in a JSON file, eg:

```json
{
  "some.key": [ "name", "age" ],
  "other.key": [ "count" ]
}
```

Then the caller could `require` the JSON file, or hardcode them, and just pass them in. Also `baseTranslations` doesn't make much sense to me - what about `knownParameters` or something?

I think the API would be a little cleaner if the optional properties were passed as an Object rather than individually named parameters. This makes it easier to include just the ones you want, and also makes it easier to add another optional parameter at a later date. eg:

```js
checkTranslations('/a/b/c', { languages: [ 'en' ], knownParameters: { ... } }).then();
```

I think rejecting the promise and letting the caller handle or ignore the errors is a good idea.

mrsarm commented:
@garethbowen cool, I'll start the new repo "translation-checker" with your suggestions and moving the reviewed code from here, and then we can continue discussing the interface details there, thanks!

mrsarm commented:
I close this PR in favor of the new implementation I'm working at https://github.com/medic/medic-conf/pull/356
